### ISSUE TITLE
Scale and Position of Objects Differ between OBJ vs DICOM/NRRD

### ISSUE THREAD

**Issue by eldrickm** [mlavik1/UnityVolumeRendering#124]
First of all, a deep thank you for working on this project. It works wonderfully and has been extremely helpful. I am planning on donating on your behalf to the ICRC.

I am working on aligning an .obj file of an object I am exporting out of 3D Slicer with it's DICOM / NRRD representation. I am exporting each of these files directly from the same data set in 3D Slicer.

However, I am observing that there are differences in the scale, rotation, and position of the origin across both imports. For example, I have exported a cuboid object from 3D slicer (Figure 1) - the peach-colored box corresponds to the OBJ file to be exported from a segmentation (Figure 2). Notice that they are basically superimposed on each other in 3D slicer.

![Figure 1](https://user-images.githubusercontent.com/13305856/196822047-60221675-2058-40fe-9ae4-28bec1b3b9c2.png)
![Figure 2](https://user-images.githubusercontent.com/13305856/196822076-1e166338-e1b2-4aad-8e4e-f2d61fb3843f.png)

When I import the OBJ file, all is as expected. (Figure 3 - I've re-oriented it so it faces up)

![Figure 3](https://user-images.githubusercontent.com/13305856/196822172-fae4324b-692d-4247-bf3f-e7af69726cb0.png)

But when I import the NRRD (DICOM is also affected) as a rendered volume (Fig. 4) I observe the following:

![Figure 4](https://user-images.githubusercontent.com/13305856/196823331-2cbbe723-5878-4760-bc87-99e5e30022ca.png)

1. There is a scale mismatch (see the dramatic size difference)
2. The position of both objects differ. The OBJ file has the correct origin relative to the object - near one of the sides, while the rendered volume's origin seems to be placed somewhere in the center of the geometry (see the difference in the coordinate axes of both objects. I am showing the coordinate axes at the pivot of both game objects). I have offset the two geometries so that it can be more easily seen.
3. Rotation seems to be unaffected - I can manually adjust only the position and scale and have the two objects align.
4. The scale and position values to correctly align the two object can be different between datasets. I've tested it with two datasets trying to see if the scale / position offsets would be constant.

Could you help me debug or find the issue? Thank you kindly in advance!

**Comment by mlavik1** [comment]
Hi @eldrickm !
Thanks for the kind words, I'm very happy to hear that :) 
When I wrote the importers I simply assumed that you would wouldn't need the relative sizes between multiple imported datasets to be right (originally you could only import one :sweat_smile: ), so I simply stored the imported dataset inside of a (1,1,1) sized box, and then scale the axes to make the the dimension right for datasets with a non-uniform dimension.
So the real size is ignored, which I realise is a big limitation.

The importers seem set the scale correctly (I think?):
- `SimpleITKImageFileImporter.cs` (for NRRD files)
- `SimpleITKImageSequenceImporter.cs` (for DICOM files)

However, the `VolumeObjectFactory.CreateObject` function then scales down the dataset to make sure all imported datasets have more or less the same size.
```csharp
if(dataset.scaleX != 0.0f && dataset.scaleY != 0.0f && dataset.scaleZ != 0.0f)
{
    float maxScale = Mathf.Max(dataset.scaleX, dataset.scaleY, dataset.scaleZ);
    volObj.transform.localScale = new Vector3(dataset.scaleX / maxScale, dataset.scaleY / maxScale, dataset.scaleZ / maxScale);
}
```
(see https://github.com/mlavik1/UnityVolumeRendering/blob/70ed1d1d94cd6544db0ec1eb7e86f6a3bcece31a/Assets/Scripts/VolumeObject/VolumeObjectFactory.cs#L45)

I suppose you could try to remove that `/ maxScale` ?
There might still be some other issues too.. And we probably need to handle different scale units? (mm, etc.)

I can also take a look at it when I have time (this weekend or early next week I think! Getting some visitors first :) )

**Comment by mlavik1** [comment]
That's great to hear! :)

So for that the relevant file to look in would be `SimpleITKImageSequenceImporter.cs`.
It doesn't seem like SimpleITK has a property for image position though.. But I think you can read DICOM attributes by tag, using the [GetMetaData](https://simpleitk.org/doxygen/latest/html/classitk_1_1simple_1_1Image.html#a43df17945f303feb16b4accf1b684515) function.
There is something example code here: https://simpleitk.readthedocs.io/en/master/link_DicomImagePrintTags_docs.html
(if you print the tags like that, you can see how they need to be formatted. I think it's "0020|0032").

Please let me know if you get that to work, or if you have any further questions or issues :)

**Comment by eldrickm** [comment]
Yes I was able to read the tags using the OpenDICOM importer - in fact it already takes it into account!

https://github.com/mlavik1/UnityVolumeRendering/blob/70ed1d1d94cd6544db0ec1eb7e86f6a3bcece31a/Assets/Scripts/Importing/ImageSequenceImporter/OpenDICOM/DICOMImporter.cs#L188

My question is how to use that metadata in positioning the volume render game object. Do you know the relationship between the coordinates that are given in that metadata field and say, the `transform.localPosition` of the volume render game object?

**Comment by mlavik1** [comment]
Oh right, you're using the OpenDICOM importer :) 
Hmm.. I guess you could use the position of the first slice then?

We sort the slices here: https://github.com/mlavik1/UnityVolumeRendering/blob/70ed1d1d94cd6544db0ec1eb7e86f6a3bcece31a/Assets/Scripts/Importing/ImageSequenceImporter/OpenDICOM/DICOMImporter.cs#L122

After that, you could fetch the location of the first slice, and maybe use that as position? (that is, `files[0].location`)

**Comment by mlavik1** [comment]
Hmm I noticed that the location is a `float` though (the X value, I suppose). 
You should probably read the X,Y,Z value as well then. The X value is read here: https://github.com/mlavik1/UnityVolumeRendering/blob/70ed1d1d94cd6544db0ec1eb7e86f6a3bcece31a/Assets/Scripts/Importing/ImageSequenceImporter/OpenDICOM/DICOMImporter.cs#L198
So simply change that to read `elemLoc.Value[0]`, `elemLoc.Value[1]`, `elemLoc.Value[2]` to a Vector3.

By the way, if you're on Windows or Linux I would recommend using the newer SimpleITK importer for DICOM btw :) https://github.com/mlavik1/UnityVolumeRendering/blob/master/Documentation/SimpleITK.md
Though, if your project is more cross platform, then I guess OpenDICOM is the only choice.

**Comment by eldrickm** [comment]
Hi Matias,

Sorry for the delay in response!

Yes, that was my thought as well, however I am not sure which game object in particular to assign the value to.
Each slice also has its own location, so I was not sure which slice's location to use either. Do you have any insight on the relationship between the slice's location numbers and the positioning of the game object as a whole?

**Comment by mlavik1** [comment]
Hi again!
Sorry for the delay. Busy week :sweat_smile: 

Hmm.. I'm not too sure about that, to be honest.
I guess the code for reading the slice locations would need to be changed a bit if you need the relative positions between the objects to be correct.

According to [this](https://dicom.innolitics.com/ciods/rt-dose/image-plane/00200032) the slice locations need to be interpreted according to another DICOM tag: Image Orientation (0020,0037)?

**Comment by eldrickm** [comment]
Thank you for the link - I'll take a closer look and will see if I can get a solution within the next few weeks!

**Comment by mlavik1** [comment]
Great, thanks! :)
